geode-dunit/src/main/java/org/apache/geode/test/dunit/rules/DistributedReference.java

```diff
@@ -187,47 +197,76 @@ public V get() {
    * @param newValue the new value
    */
   public DistributedReference<V> set(V newValue) {
-    REFERENCE.set(newValue);
+    REFERENCE.get().put(identity, newValue);
     return this;
   }
 
+  @Override
+  protected void before() {
+    invoker().invokeInEveryVMAndController(() -> invokeBefore());
+  }
+
   @Override
   protected void after() {
-    invoker().invokeInEveryVMAndController(this::invokeAfter);
+    invoker().invokeInEveryVMAndController(() -> invokeAfter());
+  }
+
+  protected void afterCreateVM(VM vm) {
+    vm.invoke(() -> invokeBefore());
+  }
+
+  protected void beforeBounceVM(VM vm) {
+    // override if needed
+  }
+
+  protected void afterBounceVM(VM vm) {
+    vm.invoke(() -> invokeBefore());
+  }
+
+  private void invokeBefore() {
+    REFERENCE.compareAndSet(null, new HashMap<>());
+    REFERENCE.get().putIfAbsent(identity, null);
   }
 
   private void invokeAfter() {
-    V value = get();
-    if (value == null) {
+    Map<Integer, Object> references = REFERENCE.getAndSet(null);
+    if (references == null) {
+      return;
+    }
+
+    for (Object object : references.values()) {
+      invokeAfter(object);
+    }
+  }
+
+  private void invokeAfter(Object object) {
+    if (object == null) {
       return;
     }
-    REFERENCE.set(null);
 
     if (autoClose.get()) {
-      autoClose(value);
+      autoClose(object);
     }
   }
 
-  private void autoClose(V value) {
-    if (value instanceof AutoCloseable) {
-      close((AutoCloseable) value);
+  private void autoClose(Object object) {
+    if (object instanceof AutoCloseable) {
+      close((AutoCloseable) object);
 
-    } else if (hasMethod(value.getClass(), "close")) {
-      invokeMethod(value, "close");
+    } else if (object instanceof AtomicBoolean) {
+      toFalse((AtomicBoolean) object);
 
-    } else if (hasMethod(value.getClass(), "disconnect")) {
-      invokeMethod(value, "disconnect");
+    } else if (object instanceof CountDownLatch) {
+      openLatch((CountDownLatch) object);
 
-    } else if (hasMethod(value.getClass(), "stop")) {
-      invokeMethod(value, "stop");
-    }
-  }
+    } else if (hasMethod(object.getClass(), "close")) {
+      invokeMethod(object, "close");
```

Comment:
Could we not leverage method overloading here.
Instead of `toFalse((AtomicBoolean)object)` we use `close(object)` and then have a `close(AtomicBoolean)` in the CompletionUtils?

Reply:
Do we really want a method called `close(AtomicBoolean)` that sets the specified value to false? I started out with multiple `close` methods and then I changed them from that to:
* toFalse(AtomicBoolean)
* openLatch(CountDownLatch)

And then what about opening `CountDownLatch`? Do we change that to `close` as well even tho close is the opposite of open?

I was thinking that my creation of `CompletionUtils` might be slightly more controversial because it's of dubious value outside `DistributedReference`.

Reply:
I imagine that the effort to write `toFalse` would be the same as `close(AtomicBoolean)`. 
I don't mind the `CompletionUtils` class. 
It is test only code anyway. I just thought the code would be cleaner with just a single `close(*)` call. Then all the logic for closing, even all the method look ups, could be contained in the `CompletionUtils` class.

Reply:
New commit includes "Rename all CompletionUtils to close". Please review again. Thanks!